### DISCUSSION TITLE
How to generate a push certificate in fastlane using AppStore connect API Key?

### DISCUSSION THREAD

**Discussion by anilsaini-fission** [fastlane/fastlane#18121] [General] [answered]
I'm using Bitrise with fastlane integration to generate 
get_push_certificate -> Command is failing 

Skipping automatic bypass of 2FA upgrade
+------------------+------------------------+
|               Lane Context                |
+------------------+------------------------+
| DEFAULT_PLATFORM | ios                    |
| PLATFORM_NAME    | ios                    |
| LANE_NAME        | ios generate_push_cert |
+------------------+------------------------+
[07:01:56]: Need to acknowledge to Apple's Apple ID and Privacy statement. Please manually log into https://appleid.apple.com (or https://appstoreconnect.apple.com) to acknowledge the statement. Your account might also be asked to upgrade to 2FA. Set SPACESHIP_SKIP_2FA_UPGRADE=1 for fastlane to automaticaly bypass 2FA upgrade if possible.
+------+----------------------------+-------------+
|                fastlane summary                 |
+------+----------------------------+-------------+
| Step | Action                     | Time (in s) |
+------+----------------------------+-------------+
| 1    | Verifying fastlane version | 0           |
| 2    | default_platform           | 0           |
| 💥   | get_push_certificate       | 0           |
+------+----------------------------+-------------+
+--------+--------------+----------------+
|        Plugin updates available        |
+--------+--------------+----------------+
| Plugin | Your Version | Latest Version |
+--------+--------------+----------------+
| sentry | 1.6.0        | 1.8.0          |
+--------+--------------+----------------+
[07:01:56]: To update all plugins, just run
[07:01:56]: $ bundle exec fastlane update_plugins
[07:01:56]: fastlane finished with errors
[!] The request could not be completed because:
	Need to acknowledge to Apple's Apple ID and Privacy statement. Please manually log into https://appleid.apple.com (or https://appstoreconnect.apple.com) to acknowledge the statement. Your account might also be asked to upgrade to 2FA. Set SPACESHIP_SKIP_2FA_UPGRADE=1 for fastlane to automaticaly bypass 2FA upgrade if possible.
Fastlane command: (bundle "_1.17.2_" "exec" "fastlane" "generate_push_cert") failed
If you want to send an issue report to fastlane (https://github.com/fastlane/fastlane/issues/new), you can find the output of fastlane env in the following log file:
/Users/vagrant/deploy/fastlane_env.log
$ bundle "_1.17.2_" "exec" "fastlane" "env"
Command failed, error: exit status 1

**Comment by dinsen** [comment]
Hi @anilsaini-fission 
Did you set `SPACESHIP_SKIP_2FA_UPGRADE=1` as the error says?

**Reply by anilsaini-fission** [reply]
Tried it however it didn't work. As explain in the https://github.com/fastlane/fastlane/issues/18098 it looks like it will work only for AppleId+Password.

**Reply by dinsen** [reply]
Correct, Apple have not added endpoints to get or create push certificates. 

Have you considered using the token key instead? You can use this for all apps on the account and it will not expire.

**Reply by anilsaini-fission** [reply]
@Dinsen - 
Did you mean App store Connect API key authentication? if so, token key don't support to push certificate yet. And token key is only applicable to Apple Organisation account.

**Reply by dinsen** [reply]
Yes I meant the API Key.

What type is your account?
Just to clarify, if you use this token/private key to send out push, you will no longer have to create push certificates.
https://help.apple.com/developer-account/#/devcdfbb56a3
https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/establishing_a_token-based_connection_to_apns

**Reply by anilsaini-fission** [reply]
Thanks @Dinsen, 
I use both organisation as well as enterprise account.

**Reply by dinsen** [reply]
It is possible to create a key for both organizations and enterprise accounts that can send out push. 
So by changing to this way of sending out push, you no longer need to create or renew any push certificates.

You create the key in the Developer portal:
https://developer.apple.com/account/resources/authkeys/list

**Comment by max-ott** [comment, marked as answer]
It's not possible to generate push certificates with the ASC API key. Apple simply has no endpoint for that. You have to use Apple ID + password for that.

**Reply by anilsaini-fission** [reply]
Thanks @max-ott
